units/event_log_tests.cc

```diff
@@ -0,0 +1,183 @@
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <iostream>
+#include <ostream>
+#include <string>
+#include <vector>
+
+#include "gtest/gtest.h"
+
+namespace {
+// Set of supported attribute types.
+enum AttributeType { String, Int64, VectorInt64 };
+// Specifies the importance of an Event. Events are logged based on their level
+// of importance. Eg., compile_time.csv only contains the important compile time
+// events while event.log contains all the compile time events.
+enum LogLevel { Low, Medium, High };
+
+std::string ValueToCSVString(const std::string &value) { return value; }
+
+std::string ValueToCSVString(const int64_t value) {
+  return std::to_string(value);
+}
+
+std::string ValueToCSVString(const std::vector<int64_t> &values) {
+  std::string tmp = "\"[";
+  using std::begin;
+  using std::end;
+  auto e = end(values);
+  for (auto it = begin(values); it != e; ++it) {
+    const char *delimiter = std::next(it) != values.end() ? "," : "";
+    tmp += std::to_string(*it) + delimiter;
+  }
+  tmp += "]\"";
+  return tmp;
+}
+
+struct Attribute {
+  const char *name;
+  AttributeType value_type;
+
+  template <typename T>
+  bool isa() {
+    return value_type == T::id;
+  }
+
+  // Safely casts the Attribute to the given type (T). Valid typenames should be
+  // derived from Attribute and have an AttributeType `id` field.
+  template <typename T>
+  T *cast() {
+    return isa<T>() ? static_cast<T *>(this) : nullptr;
+  };
+};
+
+// Implements the Attribute base class for each specific `AttributeType`.
+// Valid typenames: {std::string, int64_t, std::vector<int64_t>}
+// Example:
+//    AttributeImpl<int64_t, AttributeType::Int64> timestamp("timestamp", 1234);
+template <typename T, AttributeType VT>
+struct AttributeImpl : Attribute {
+  T value;
+  static constexpr AttributeType id = VT;
+
+  AttributeImpl(const char *_name, T _value)
```

Comment:
Because this code is generic and will handle containers like `vector` or `string`, we should avoid needless copying.   We should take the value by a const reference.